src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,449 @@
+
+ private:
+  LoweringToDatalogPass() : fresh_var_count_(0) {}
+
+  // This returns a fresh variable name. Fresh variables are introduced in a
+  // few places during the translation.
+  std::string FreshVar() {
+    assert(fresh_var_count_ < std::numeric_limits<uint64_t>::max());
+    return "x__" + std::to_string(++fresh_var_count_);
+  }
+
+  // This function makes a copy of a predicate that prepends the name with
+  // another string and prepends the list of arguments with additional
+  // arguments. This is used in a few places in the translation, for
+  // example, to translate "X says blah(args)" into "says_blah(X, args)".
+  Predicate PushOntoPred(std::string modifier,
+                         std::vector<std::string> new_args,
+                         Predicate predicate) {
+    std::string new_name = modifier + predicate.name();
+    new_args.insert(new_args.end(),
+                    std::make_move_iterator(predicate.args().begin()),
+                    std::make_move_iterator(predicate.args().end()));
+    Sign sign_copy = predicate.sign();
+    return Predicate(new_name, new_args, sign_copy);
+  }
+
+  Predicate PushPrin(std::string modifier, Principal principal,
+                     Predicate predicate) {
+    return PushOntoPred(modifier, {principal.name()}, predicate);
+  }
+
+  Predicate AttributeToDLIR(Attribute attribute) {
+    // If attribute is `X pred(args...)` the following predicate is
+    // `pred(X, args...)`
+    return PushPrin(std::string(""), attribute.principal(),
+                    attribute.predicate());
+  }
+
+  // Translating an attribute results in two objects:
+  // - A predicate, this is the main result of the translation
+  // - An additional rule that allows attributes to be passed with canActAs
+  // This function is for generating the additional rule.
+  // This takes both a speaker and the actual attribute as argument because
+  // the speaker is needed to generate the additional rule. Note that all base
+  // facts have a speaker either explicitly or implicitly:
+  // - If it appears on the LHS of an assertion, it explicitly has a speaker
+  // - If it appears on the RHS of an assertion, it behaves semantically
+  // like it has the same speaker as the head of the assertion.
+  DLIRAssertion SpokenAttributeToDLIR(Principal speaker, Attribute attribute) {
+    Predicate main_predicate = AttributeToDLIR(attribute);
+
+    // Attributes interact with "canActAs" because if "Y canActAs X"
+    // then Y also picks up X's attributes. We need to generate
+    // an additional rule to implement this behavior. If the attribute
+    // under translation is `X PredX`, the additional rule is:
+    // `speaker says Y PredX :-
+    //    speaker says Y canActAs X, speaker says X PredX`
+    // (Where Y is a fresh variable)
+    Principal prin_y = Principal(FreshVar());
+
+    // This is `speaker says Y PredX`
+    Predicate generated_lhs =
+        PushPrin(std::string("says_"), prin_y, main_predicate);
```

Comment:
In general, if you're wrapping a string literal with `std::string`, that's a sign that something is suboptimal. `std::string` is basically an owned buffer of characters, and introduces a copy (which the compiler may or may not be able to magically eliminate through optimizations). This is bad because string literals are highly optimized; each string literal within a translation unit with the same contents points to the same place in the `.rodata` section. Usually, this means that the parameter that this argument feeds into should be an `absl::string_view`, which handles this case without copying.

Reply:
is `absl::string_view` better than just leaving it as a literal? I just left it as a literal.